### ISSUE TITLE
是否支持图像CNN算子操作

### ISSUE THREAD

**Issue by maoweinuaa** [FederatedAI/FATE#1488]
我按照官方的例子，将minst的网络结构改成带有卷积算子的网络之后，发现无法训练。请问fate目前是否支持这块？感谢

**Comment by weiwee** [comment]
> 我按照官方的例子，将minst的网络结构改成带有卷积算子的网络之后，发现无法训练。请问fate目前是否支持这块？感谢

算是支持的，但是需要做一些转换
希望提供如下信息：
1. 系统版本
2.详细配置
3. 相关错误

**Comment by maoweinuaa** [comment]
> > 我按照官方的例子，将minst的网络结构改成带有卷积算子的网络之后，发现无法训练。请问fate目前是否支持这块？感谢
> 
> 算是支持的，但是需要做一些转换
> 希望提供如下信息：
> 
> 1. 系统版本
> 2. 详细配置
> 3. 相关错误
版本1.4的。用的kubefate在2个机器上部署的，我之前用没有卷积的可以运行的，就是，修改了网络就不行了。老师能否加下你微信

**Comment by weiwee** [comment]
> > > 我按照官方的例子，将minst的网络结构改成带有卷积算子的网络之后，发现无法训练。请问fate目前是否支持这块？感谢
> > 
> > 
> > 算是支持的，但是需要做一些转换
> > 希望提供如下信息：
> > 
> > 1. 系统版本
> > 2. 详细配置
> > 3. 相关错误
> >    版本1.4的。用的kubefate在2个机器上部署的，我之前用没有卷积的可以运行的，就是，修改了网络就不行了。老师能否加下你微信
可以参考尝试这个脚本

from string import Template


def build_model():
    from keras.layers import Conv2D, MaxPooling2D
    from keras.layers import Dense, Dropout, Flatten, Reshape
    from keras.models import Sequential

    model = Sequential()
    model.add(Reshape((28, 28)))
    model.add(Conv2D(32, kernel_size=(3, 3), activation='relu', input_shape=(28, 28, 1)))
    model.add(Conv2D(64, (3, 3), activation='relu'))
    model.add(MaxPooling2D(pool_size=(2, 2)))
    model.add(Dropout(0.25))
    model.add(Flatten())
    model.add(Dense(128, activation='relu'))
    model.add(Dropout(0.5))
    model.add(Dense(10, activation='softmax'))
    return model


def save_runtime_conf(model):
    nn_define = model.to_json()
    temp = open("test_homo_nn_keras_temperate.json").read()
    json = Template(temp).substitute(nn_define=nn_define)

    with open("test_homo_nn_keras.json", "w") as f:
        f.write(json)


if __name__ == '__main__':
    model = build_model()
    save_runtime_conf(model)

这个脚本在homo nn的example下去生成conf

**Comment by maoweinuaa** [comment]
对的，就是用的这个，跑不起来的

**Comment by maoweinuaa** [comment]
> > > > 我按照官方的例子，将minst的网络结构改成带有卷积算子的网络之后，发现无法训练。请问fate目前是否支持这块？感谢
> > > 
> > > 
> > > 算是支持的，但是需要做一些转换
> > > 希望提供如下信息：
> > > 
> > > 1. 系统版本
> > > 2. 详细配置
> > > 3. 相关错误
> > >    版本1.4的。用的kubefate在2个机器上部署的，我之前用没有卷积的可以运行的，就是，修改了网络就不行了。老师能否加下你微信
> 
> 可以参考尝试这个脚本
> 
> ```
> from string import Template
> 
> 
> def build_model():
>     from keras.layers import Conv2D, MaxPooling2D
>     from keras.layers import Dense, Dropout, Flatten, Reshape
>     from keras.models import Sequential
> 
>     model = Sequential()
>     model.add(Reshape((28, 28)))
>     model.add(Conv2D(32, kernel_size=(3, 3), activation='relu', input_shape=(28, 28, 1)))
>     model.add(Conv2D(64, (3, 3), activation='relu'))
>     model.add(MaxPooling2D(pool_size=(2, 2)))
>     model.add(Dropout(0.25))
>     model.add(Flatten())
>     model.add(Dense(128, activation='relu'))
>     model.add(Dropout(0.5))
>     model.add(Dense(10, activation='softmax'))
>     return model
> 
> 
> def save_runtime_conf(model):
>     nn_define = model.to_json()
>     temp = open("test_homo_nn_keras_temperate.json").read()
>     json = Template(temp).substitute(nn_define=nn_define)
> 
>     with open("test_homo_nn_keras.json", "w") as f:
>         f.write(json)
> 
> 
> if __name__ == '__main__':
>     model = build_model()
>     save_runtime_conf(model)
> ```
> 
> 这个脚本在homo nn的example下去生成conf

老师我的问题能解决吗？

**Comment by weiwee** [comment]
> > > > > 我按照官方的例子，将minst的网络结构改成带有卷积算子的网络之后，发现无法训练。请问fate目前是否支持这块？感谢
> > > > 
> > > > 
> > > > 算是支持的，但是需要做一些转换
> > > > 希望提供如下信息：
> > > > 
> > > > 1. 系统版本
> > > > 2. 详细配置
> > > > 3. 相关错误
> > > >    版本1.4的。用的kubefate在2个机器上部署的，我之前用没有卷积的可以运行的，就是，修改了网络就不行了。老师能否加下你微信
> > 
> > 
> > 可以参考尝试这个脚本
> > ```
> > from string import Template
> > 
> > 
> > def build_model():
> >     from keras.layers import Conv2D, MaxPooling2D
> >     from keras.layers import Dense, Dropout, Flatten, Reshape
> >     from keras.models import Sequential
> > 
> >     model = Sequential()
> >     model.add(Reshape((28, 28)))
> >     model.add(Conv2D(32, kernel_size=(3, 3), activation='relu', input_shape=(28, 28, 1)))
> >     model.add(Conv2D(64, (3, 3), activation='relu'))
> >     model.add(MaxPooling2D(pool_size=(2, 2)))
> >     model.add(Dropout(0.25))
> >     model.add(Flatten())
> >     model.add(Dense(128, activation='relu'))
> >     model.add(Dropout(0.5))
> >     model.add(Dense(10, activation='softmax'))
> >     return model
> > 
> > 
> > def save_runtime_conf(model):
> >     nn_define = model.to_json()
> >     temp = open("test_homo_nn_keras_temperate.json").read()
> >     json = Template(temp).substitute(nn_define=nn_define)
> > 
> >     with open("test_homo_nn_keras.json", "w") as f:
> >         f.write(json)
> > 
> > 
> > if __name__ == '__main__':
> >     model = build_model()
> >     save_runtime_conf(model)
> > ```
> > 
> > 
> > 这个脚本在homo nn的example下去生成conf
> 
> 老师我的问题能解决吗？

我提了个pr，包含了详细的mnist demo(homo_nn/mnist_demo目录下)， 我自己验证过是可以跑的，但是效果还没调试，你可以参考修改你的版本，希望能帮到你

**Comment by b02202050** [comment]
请问各位前辈，既然FATE支援CNN算子，那么FATE支援GPU运算吗？

**Comment by weiwee** [comment]
> 请问各位前辈，既然FATE支持CNN算子，那么FATE支持GPU运算吗？

支持，安装好cuda/cudnn/tensorflow-gpu==1.15.4, nn会自动使用可用的gpu资源

**Comment by weiwee** [comment]
> > > > > > 我按照官方的例子，将minst的网络结构改成带有卷积算子的网络之后，发现无法训练。请问fate目前是否支持这块？感谢
> > > > > 
> > > > > 
> > > > > 算是支持的，但是需要做一些转换
> > > > > 希望提供如下信息：
> > > > > 
> > > > > 1. 系统版本
> > > > > 2. 详细配置
> > > > > 3. 相关错误
> > > > >    版本1.4的。用的kubefate在2个机器上部署的，我之前用没有卷积的可以运行的，就是，修改了网络就不行了。老师能否加下你微信
> > > 
> > > 
> > > 可以参考尝试这个脚本
> > > ```
> > > from string import Template
> > > 
> > > 
> > > def build_model():
> > >     from keras.layers import Conv2D, MaxPooling2D
> > >     from keras.layers import Dense, Dropout, Flatten, Reshape
> > >     from keras.models import Sequential
> > > 
> > >     model = Sequential()
> > >     model.add(Reshape((28, 28)))
> > >     model.add(Conv2D(32, kernel_size=(3, 3), activation='relu', input_shape=(28, 28, 1)))
> > >     model.add(Conv2D(64, (3, 3), activation='relu'))
> > >     model.add(MaxPooling2D(pool_size=(2, 2)))
> > >     model.add(Dropout(0.25))
> > >     model.add(Flatten())
> > >     model.add(Dense(128, activation='relu'))
> > >     model.add(Dropout(0.5))
> > >     model.add(Dense(10, activation='softmax'))
> > >     return model
> > > 
> > > 
> > > def save_runtime_conf(model):
> > >     nn_define = model.to_json()
> > >     temp = open("test_homo_nn_keras_temperate.json").read()
> > >     json = Template(temp).substitute(nn_define=nn_define)
> > > 
> > >     with open("test_homo_nn_keras.json", "w") as f:
> > >         f.write(json)
> > > 
> > > 
> > > if __name__ == '__main__':
> > >     model = build_model()
> > >     save_runtime_conf(model)
> > > ```
> > > 
> > > 
> > > 这个脚本在homo nn的example下去生成conf
> > 
> > 
> > 老师我的问题能解决吗？
> 
> 我提了个pr(#1495)，包含了详细的mnist demo(homo_nn/mnist_demo目录下)， 我自己验证过是可以跑的，但是效果还没调试，你可以参考修改你的版本，希望能帮到你

1.5版本提供了功能跟效果都ok的mnist demo，在dsl/v1/homo_nn下可以找到

**Comment by b02202050** [comment]
謝謝！
想再請問如果使用PyTorch的模型支援GPU嗎？
